File: sim/world/locations.py
"""
world/locations.py
All named locations in Barangay Mabuhay.
Matches the map zones from the frontend.
"""

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class Location:
    key: str
    label: str
    zone_type: str          # government | commercial | civic | residential | agricultural | industrial
    capacity: int           # max agents before it feels 'crowded'
    x: float                # normalized 0-1 (matches frontend)
    y: float
    w: float
    h: float
    current_agents: List[int] = field(default_factory=list)   # agent ids present

    def is_crowded(self) -> bool:
        return len(self.current_agents) > self.capacity

    def add_agent(self, agent_id: int):
        if agent_id not in self.current_agents:
            self.current_agents.append(agent_id)

    def remove_agent(self, agent_id: int):
        if agent_id in self.current_agents:
            self.current_agents.remove(agent_id)


LOCATIONS: dict[str, Location] = {
    "farm":       Location("farm",       "Farm",              "agricultural", 80,  .18, .32, .14, .16),
    "factory":    Location("factory",    "Factories",         "industrial",   120, .46, .10, .22, .16),
    "church":     Location("church",     "Church",            "civic",        60,  .44, .38, .08, .07),
    "school":     Location("school",     "School",            "civic",        200, .38, .42, .09, .07),
    "hospital":   Location("hospital",   "Hospital",          "civic",        80,  .56, .38, .11, .07),
    "park":       Location("park",       "Park",              "civic",        150, .40, .50, .08, .07),
    "media":      Location("media",      "Media Station",     "civic",        20,  .54, .48, .08, .07),
    "bgyhall":    Location("bgyhall",    "Barangay Hall",     "government",   40,  .64, .48, .08, .07),
    "market":     Location("market",     "Market",            "commercial",   300, .42, .58, .10, .07),
    "police":     Location("police",     "Police Station",    "government",   30,  .54, .57, .07, .06),
    "audit":      Location("audit",      "Audit Office",      "government",   15,  .63, .57, .06, .06),
    "sarisari":   Location("sarisari",   "Sari-Sari Stores",  "commercial",   50,  .34, .62, .07, .07),
    "contoffice": Location("contoffice", "Contractor Office", "commercial",   20,  .38, .68, .07, .06),
    "bank":       Location("bank",       "Bank",              "commercial",   40,  .46, .68, .07, .06),
    "munhall":    Location("munhall",    "Municipal Hall",    "government",   60,  .54, .63, .12, .09),
    "resA":       Location("resA",       "Residential A",     "residential",  600, .74, .55, .14, .18),
    "resB":       Location("resB",       "Residential B",     "residential",  1600,.42, .80, .20, .12),
    "resC":       Location("resC",       "Residential C",     "residential",  2800,.12, .55, .22, .22),
}


def get_location(key: str) -> Optional[Location]:
    return LOCATIONS.get(key)


def government_locations() -> List[str]:
    return [k for k, v in LOCATIONS.items() if v.zone_type == "government"]


def civic_locations() -> List[str]:
    return [k for k, v in LOCATIONS.items() if v.zone_type == "civic"]

File: frontend/lib/locations.ts
// Zone definitions mirroring sim/world/locations.py

export type ZoneType = "government" | "commercial" | "civic" | "residential" | "agricultural" | "industrial";

export interface ZoneDefinition {
  key: string;
  label: string;
  zone_type: ZoneType;
  x: number;
  y: number;
  w: number;
  h: number;
  fillColor: string;
  strokeColor: string;
}

export const ZONES: Record<string, ZoneDefinition> = {
  farm:       { key: "farm",       label: "FARM",           zone_type: "agricultural", x: 0.18, y: 0.32, w: 0.14, h: 0.16, fillColor: "#5a3a1a", strokeColor: "#3a2208" },
  factory:    { key: "factory",    label: "FACTORIES",      zone_type: "industrial",   x: 0.46, y: 0.10, w: 0.22, h: 0.16, fillColor: "#444",    strokeColor: "#333"    },
  church:     { key: "church",     label: "CHURCH",         zone_type: "civic",        x: 0.44, y: 0.38, w: 0.08, h: 0.07, fillColor: "#1a5c1a", strokeColor: "#0f3a0f" },
  school:     { key: "school",     label: "SCHOOL",         zone_type: "civic",        x: 0.38, y: 0.42, w: 0.09, h: 0.07, fillColor: "#1a5c1a", strokeColor: "#0f3a0f" },
  hospital:   { key: "hospital",   label: "HOSPITAL",       zone_type: "civic",        x: 0.56, y: 0.38, w: 0.11, h: 0.07, fillColor: "#1a5c1a", strokeColor: "#0f3a0f" },
  park:       { key: "park",       label: "PARK",           zone_type: "civic",        x: 0.40, y: 0.50, w: 0.08, h: 0.07, fillColor: "#1a5c1a", strokeColor: "#0f3a0f" },
  media:      { key: "media",      label: "MEDIA",          zone_type: "civic",        x: 0.54, y: 0.48, w: 0.08, h: 0.07, fillColor: "#1a5c1a", strokeColor: "#0f3a0f" },
  bgyhall:    { key: "bgyhall",    label: "BRGY HALL",      zone_type: "government",   x: 0.64, y: 0.48, w: 0.08, h: 0.07, fillColor: "#8b1a1a", strokeColor: "#6b0f0f" },
  market:     { key: "market",     label: "MARKET",         zone_type: "commercial",   x: 0.42, y: 0.58, w: 0.10, h: 0.07, fillColor: "#7a5230", strokeColor: "#5a3a1a" },
  police:     { key: "police",     label: "POLICE",         zone_type: "government",   x: 0.54, y: 0.57, w: 0.07, h: 0.06, fillColor: "#8b1a1a", strokeColor: "#6b0f0f" },
  audit:      { key: "audit",      label: "AUDIT",          zone_type: "government",   x: 0.63, y: 0.57, w: 0.06, h: 0.06, fillColor: "#8b1a1a", strokeColor: "#6b0f0f" },
  sarisari:   { key: "sarisari",   label: "SARI-SARI",      zone_type: "commercial",   x: 0.34, y: 0.62, w: 0.07, h: 0.07, fillColor: "#7a5230", strokeColor: "#5a3a1a" },
  contoffice: { key: "contoffice", label: "CONTRACTOR",     zone_type: "commercial",   x: 0.38, y: 0.68, w: 0.07, h: 0.06, fillColor: "#7a5230", strokeColor: "#5a3a1a" },
  bank:       { key: "bank",       label: "BANK",           zone_type: "commercial",   x: 0.46, y: 0.68, w: 0.07, h: 0.06, fillColor: "#7a5230", strokeColor: "#5a3a1a" },
  munhall:    { key: "munhall",    label: "MUNICIPAL HALL", zone_type: "government",   x: 0.54, y: 0.63, w: 0.12, h: 0.09, fillColor: "#8b1a1a", strokeColor: "#6b0f0f" },
  resA:       { key: "resA",       label: "RES. A",         zone_type: "residential",  x: 0.74, y: 0.55, w: 0.14, h: 0.18, fillColor: "#1e4d8c", strokeColor: "#1a3a6b" },
  resB:       { key: "resB",       label: "RES. B",         zone_type: "residential",  x: 0.42, y: 0.80, w: 0.20, h: 0.12, fillColor: "#7a2a8c", strokeColor: "#5a1a6b" },
  resC:       { key: "resC",       label: "RES. C",         zone_type: "residential",  x: 0.12, y: 0.55, w: 0.22, h: 0.22, fillColor: "#b36000", strokeColor: "#8b4a00" },
};

export const ZONE_LIST = Object.values(ZONES);

export function getZone(key: string): ZoneDefinition | undefined {
  return ZONES[key];
}

export function zoneCenter(zone: ZoneDefinition, canvasW: number, canvasH: number): { x: number; y: number } {
  return {
    x: (zone.x + zone.w / 2) * canvasW,
    y: (zone.y + zone.h / 2) * canvasH,
  };
}
